packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectInternal.kt

```diff
@@ -43,130 +31,36 @@ import kotlin.reflect.KClass
-        // We should never get here unless it is a managed object as unmanaged doesn't support observing
-        return RealmInterop.realm_object_add_notification_callback(this.`$realm$ObjectPointer`!!, callback)
-    }
-
-    override fun emitFrozenUpdate(
-        frozenRealm: RealmReference,
-        change: NativePointer,
-        channel: SendChannel<ObjectChange<RealmObjectInternal>>
-    ): ChannelResult<Unit>? {
-        val frozenObject: RealmObjectInternal? = this.freeze(frozenRealm)
-
-        return if (frozenObject == null) {
-            channel
-                .trySend(DeletedObjectImpl())
-                .also {
-                    channel.close()
-                }
-        } else {
-            val changedFieldNames = getChangedFieldNames(frozenRealm, change)
-
-            // We can identify the initial ObjectChange event emitted by core because it has no changed fields.
-            if (changedFieldNames.isEmpty()) {
-                channel.trySend(InitialObjectImpl(frozenObject))
-            } else {
-                channel.trySend(UpdatedObjectImpl(frozenObject, changedFieldNames))
-            }
-        }
-    }
-
-    private fun getChangedFieldNames(
-        frozenRealm: RealmReference,
-        change: NativePointer
-    ): Array<String> {
-        return RealmInterop.realm_object_changes_get_modified_properties(
-            change
-        ).map { propertyKey: PropertyKey ->
-            `$realm$metadata`?.get(propertyKey)?.name ?: ""
-        }.toTypedArray()
-    }
-
-    override fun asFlow(): Flow<ObjectChange<RealmObjectInternal>> {
-        return this.`$realm$Owner`!!.owner.registerObserver(this)
-    }
+public interface RealmObjectInternal : RealmObject {
+    public var `$realm$objectReference`: ObjectReference<out RealmObject>? // RealmObjectReference ?
+}
 
-    override fun delete() {
-        if (isFrozen()) {
-            throw IllegalArgumentException(
-                "Frozen objects cannot be deleted. They must be converted to live objects first " +
-                    "by using `MutableRealm/DynamicMutableRealm.findLatest(frozenObject)`."
-            )
-        }
-        if (!isValid()) {
-            throw IllegalArgumentException("Cannot perform this operation on an invalid/deleted object")
-        }
-        `$realm$ObjectPointer`?.let { RealmInterop.realm_object_delete(it) }
-    }
+internal fun RealmObject.asObjectReference(): ObjectReference<out RealmObject>? {
+    return (this as RealmObjectInternal).`$realm$objectReference`
 }
 
-internal fun RealmObject.realmObjectInternal(): RealmObjectInternal {
-    return this as RealmObjectInternal
+internal fun RealmObject.isManaged(): Boolean {
+    return asObjectReference() != null
 }
 
 internal fun RealmObjectInternal.checkValid() {
     if (!this.isValid()) {
         throw IllegalStateException("Cannot perform this operation on an invalid/deleted object")
     }
 }
+
+internal fun RealmObjectInternal.propertyInfoOrThrow(
```

Comment:
Why do we need those? My thought was that we should guard all entry points with `isManaged` (or something like that) and then only work with `ObjectReference`s internally. Maybe by a pattern like:
```
realmObject.asObjectRefernce?.<method> {
   ...
} ?: throw Illegal ...
```
Maybe I need to check it out and play with it 🤔